components/script/dom/idbopendbrequest.rs

```diff
@@ -0,0 +1,371 @@
+                        // If an upgrade event was created, it will be responsible for
+                        // dispatching the success event
+                        if did_upgrade {
+                            return;
+                        }
+                        let request = trusted_request.root();
+                        let global = request.global();
+                        match result {
+                            Ok(db) => {
+                                request.dispatch_success(&*db);
+                            },
+                            Err(dom_exception) => {
+                                request.set_result(HandleValue::undefined().into());
+                                request.set_error(dom_exception, CanGc::note());
+                                let event = Event::new(
+                                    &global,
+                                    Atom::from("error"),
+                                    EventBubbles::Bubbles,
+                                    EventCancelable::Cancelable,
+                                    CanGc::note()
+                                );
+                                event.fire(request.upcast(), CanGc::note());
+                            }
+                        }
+                    }),
+                );
+            }),
+        );
+
+        global
+            .resource_threads()
+            .sender()
+            .send(IndexedDBThreadMsg::Sync(open_operation))
+            .unwrap();
+    }
+
+    pub fn delete_database(&self, name: String) {
+        let global = self.global();
+
+        let (sender, receiver) = ipc::channel(global.time_profiler_chan().clone()).unwrap();
+        let task_source = global
+            .task_manager()
+            .database_access_task_source()
+            .to_sendable();
+        let response_listener = OpenRequestListener {
+            open_request: Trusted::new(self),
+        };
+
+        let delete_operation =
+            SyncOperation::DeleteDatabase(sender, global.origin().immutable().clone(), name);
+
+        ROUTER.add_typed_route(
+            receiver.to_ipc_receiver(),
+            Box::new(move |message| {
+                let response_listener = response_listener.clone();
+                task_source.queue(task!(request_callback: move || {
+                    response_listener.handle_delete_db(message.unwrap(), CanGc::note());
+                }));
+            }),
+        );
+
+        global
+            .resource_threads()
+            .sender()
+            .send(IndexedDBThreadMsg::Sync(delete_operation))
+            .unwrap();
+    }
+
+    #[allow(unsafe_code)]
+    pub fn dispatch_success(&self, result: &IDBDatabase) {
+        let global = self.global();
+        let this = Trusted::new(self);
+        let result = Trusted::new(result);
+
+        global.task_manager().database_access_task_source().queue(
```

Comment:
```suggestion
        // TODO: some callers expect this to be synchronous, rather than an async task.
        global.task_manager().database_access_task_source().queue(
```